### PR TITLE
health_checker: Make health check loop wait for any required SDS secrets to be loaded before starting.

### LINKED ISSUE #12389: Should HC activation be delayed until needed secrets are available?
I did a cursory walk through the code in charge of activating HCs and I don't think it waits for secrets to be ready.

Furthermore, the `initial_fetch_timeout` broadly refers to the config gRPC stream becoming active (not necessarily to the actual payloads being received):

https://github.com/envoyproxy/envoy/blob/master/docs/root/intro/arch_overview/operations/init.rst

So for clusters where we don't set an `initial_jitter` we hit `upstream_context_secrets_not_ready` during hot restarts and we also see failed upstream health checks. I didn't test this for cold starts -- though it might be an issue there too.

I think we should either:

a) not activate HCs until the needed secrets are available

or

b) document the async nature of SDS and how setting `initial_jitter` might be necessary to workaround a race between the first HCs and secrets becoming available (we could add this to https://github.com/envoyproxy/envoy/blob/master/docs/root/intro/arch_overview/operations/init.rst
)

Thoughts?

cc: @mattklein123 @fishcakez

### PR DESCRIPTION
**Pull request by mpuncel**
Commit Message: Make health check loop wait for any required SDS secrets to be loaded before starting.
Additional Description: This should avoid an issue where Envoy might take over a minute to warm clusters with default settings when SDS and active health checking are used. The issue was caused by health checks starting before secrets are ready and then waiting for `no_traffic_interval` (default 60s) before trying again. This change makes health checks wait for SDS secrets to be ready before starting. 
Risk Level: Medium
Testing: Concurrency test covers the bug that caused this to be reverted initially, and was tested locally with --runs_per_test 100, confirming test failures when the bug is present and none when it was fixed.
Docs Changes: None
Release Notes: Included
Platform Specific Features: None
Fixes #12389, #15977

### CODE REVIEW

**Comment by mpuncel** [comment]
For reviewers: the first commit is basically just a revert of https://github.com/envoyproxy/envoy/pull/13516 and then rebased onto `main`. It still contains the bug that caused it to be reverted.

The second commit has the fix, and then ClientAddSecretsReadyCallbackParallel and ServerAddSecretsReadyCallbackParallel tests I added were able to reproduce the problem a small percentage of the time.

Run on my local machine demonstrating the bug:
```
vscode@3781f2aa0124:/workspaces/envoy$ bazel test //test/extensions/transport_sockets/tls:ssl_socket_test  --test_env=ENVOY_IP_TEST_VERSIONS=v4only  --test_filter=IpVersions/SslSocketTest.ClientAddSecretsReadyCallbackParallel/IPv4 --cache_test_results=no --runs_per_test=100
INFO: Build option --test_filter has changed, discarding analysis cache.
DEBUG: /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/external/bazel_toolchains/rules/rbe_repo.bzl:491:10: Bazel 3.7.2 is used in rbe_ubuntu_clang.
INFO: Analyzed target //test/extensions/transport_sockets/tls:ssl_socket_test (0 packages loaded, 15850 targets configured).
INFO: Found 1 test target...
FAIL: //test/extensions/transport_sockets/tls:ssl_socket_test (shard 1 of 4, run 67 of 100) (see /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_67_of_100/test.log)
FAIL: //test/extensions/transport_sockets/tls:ssl_socket_test (shard 1 of 4, run 60 of 100) (see /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_60_of_100/test.log)
FAIL: //test/extensions/transport_sockets/tls:ssl_socket_test (shard 1 of 4, run 41 of 100) (see /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_41_of_100/test.log)
FAIL: //test/extensions/transport_sockets/tls:ssl_socket_test (shard 1 of 4, run 27 of 100) (see /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_27_of_100/test.log)
FAIL: //test/extensions/transport_sockets/tls:ssl_socket_test (shard 1 of 4, run 16 of 100) (see /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_16_of_100/test.log)
FAIL: //test/extensions/transport_sockets/tls:ssl_socket_test (shard 1 of 4, run 6 of 100) (see /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_6_of_100/test.log)
FAIL: //test/extensions/transport_sockets/tls:ssl_socket_test (shard 1 of 4, run 2 of 100) (see /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_2_of_100/test.log)

FAILED: //test/extensions/transport_sockets/tls:ssl_socket_test (Summary)
      /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_67_of_100/test.log
      /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_60_of_100/test.log
      /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_41_of_100/test.log
      /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_27_of_100/test.log
      /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_16_of_100/test.log
      /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_6_of_100/test.log
      /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_2_of_100/test.log
Target //test/extensions/transport_sockets/tls:ssl_socket_test up-to-date:
  bazel-bin/test/extensions/transport_sockets/tls/ssl_socket_test
INFO: Elapsed time: 24.254s, Critical Path: 1.12s
INFO: 401 processes: 1 internal, 400 processwrapper-sandbox.
INFO: Build completed, 1 test FAILED, 401 total actions
//test/extensions/transport_sockets/tls:ssl_socket_test                  FAILED in 7 out of 400 in 0.9s
  Stats over 400 runs: max = 0.9s, min = 0.3s, avg = 0.4s, dev = 0.1s
  /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_67_of_100/test.log
  /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_60_of_100/test.log
  /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_41_of_100/test.log
  /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_27_of_100/test.log
  /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_16_of_100/test.log
  /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_6_of_100/test.log
  /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/extensions/transport_sockets/tls/ssl_socket_test/shard_1_of_4_run_2_of_100/test.log

Executed 1 out of 1 test: 1 fails locally.
INFO: Build completed, 1 test FAILED, 401 total actions
```

And after it was fixed:
```
vscode@3781f2aa0124:/workspaces/envoy$ bazel test //test/extensions/transport_sockets/tls:ssl_socket_test  --test_env=ENVOY_IP_TEST_VERSIONS=v4only  --test_filter=IpVersions/SslSocketTest.ClientAddSecretsReadyCallbackParallel/IPv4 --cache_test_results=no --runs_per_test=100
DEBUG: /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/external/bazel_toolchains/rules/rbe_repo.bzl:491:10: Bazel 3.7.2 is used in rbe_ubuntu_clang.
INFO: Analyzed target //test/extensions/transport_sockets/tls:ssl_socket_test (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
Target //test/extensions/transport_sockets/tls:ssl_socket_test up-to-date:
  bazel-bin/test/extensions/transport_sockets/tls/ssl_socket_test
INFO: Elapsed time: 41.045s, Critical Path: 15.09s
INFO: 403 processes: 1 internal, 402 processwrapper-sandbox.
INFO: Build completed successfully, 403 total actions
//test/extensions/transport_sockets/tls:ssl_socket_test                  PASSED in 0.8s
  Stats over 400 runs: max = 0.8s, min = 0.3s, avg = 0.4s, dev = 0.1s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 403 total actions
vscode@3781f2aa0124:/workspaces/envoy$ bazel test //test/extensions/transport_sockets/tls:ssl_socket_test  --test_env=ENVOY_IP_TEST_VERSIONS=v4only  --test_filter=IpVersions/SslSocketTest.ServerAddSecretsReadyCallbackParallel/IPv4 --cache_test_results=no --runs_per_test=100
INFO: Build option --test_filter has changed, discarding analysis cache.
DEBUG: /build/.cache/bazel/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/external/bazel_toolchains/rules/rbe_repo.bzl:491:10: Bazel 3.7.2 is used in rbe_ubuntu_clang.
INFO: Analyzed target //test/extensions/transport_sockets/tls:ssl_socket_test (0 packages loaded, 15850 targets configured).
INFO: Found 1 test target...
Target //test/extensions/transport_sockets/tls:ssl_socket_test up-to-date:
  bazel-bin/test/extensions/transport_sockets/tls/ssl_socket_test
INFO: Elapsed time: 21.857s, Critical Path: 0.68s
INFO: 401 processes: 1 internal, 400 processwrapper-sandbox.
INFO: Build completed successfully, 401 total actions
//test/extensions/transport_sockets/tls:ssl_socket_test                  PASSED in 0.5s
  Stats over 400 runs: max = 0.5s, min = 0.3s, avg = 0.3s, dev = 0.0s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 401 total actions
```

File: test/extensions/transport_sockets/tls/ssl_socket_test.cc [hunk 1]
```diff
@@ -4756,6 +4769,188 @@ TEST_P(SslSocketTest, UpstreamNotReadySslSocket) {
...
+  MockFunction<void()> second_callback_;
+  EXPECT_CALL(second_callback_, Call());
+  server_ssl_socket_factory.addReadyCb(second_callback_.AsStdFunction());
+}
+
+// Tests adding a callback and adding secrets in parallel.  This is intended to
+// catch a race condition introduced in
+// https://github.com/envoyproxy/envoy/pull/13516 where a callback would never
+// be called due to a concurrency bug.
+TEST_P(SslSocketTest, ServerAddSecretsReadyCallbackParallel) {
```
Comment (mpuncel) [inline test/extensions/transport_sockets/tls/ssl_socket_test.cc:4867]: This and the next test are probably a bit odd since they start a thread to try to tickle a concurrency bug that has been fixed now, and only happened in ~5% of runs on my machine. This means if there is a regression for some reason we might not see it in the PR that introduces it.

Let me know if there is a better way of handling tests like this

File: test/extensions/transport_sockets/tls/ssl_socket_test.cc [hunk 2]
```diff
@@ -4756,6 +4769,188 @@ TEST_P(SslSocketTest, UpstreamNotReadySslSocket) {
...
+
+  // Add a callback in a thread to potentially tickle the concurrency bug.
+  std::thread t([&server_ssl_socket_factory, &mock_callback_]() {
+    server_ssl_socket_factory.addReadyCb(mock_callback_.AsStdFunction());
+  });
+
+  // 1 microsecond of sleep seems to be a sweet spot for giving the addReadyCb
+  // thread enough of a head start to reliably cause the test to fail with buggy
+  // code.
+  std::this_thread::sleep_for(std::chrono::microseconds(1));
```
Comment (mpuncel) [inline test/extensions/transport_sockets/tls/ssl_socket_test.cc:4904]: I'm failing the style check because of this, I'm looking into whether the issue can be reproduced with ThreadSynchronizer instead.

It's a bit weird because with the buggy code I certainly can reproduce it 100% of the time with that, but then when it is fixed there is nowhere I can put the syncPoint() that would let me deadlock it (because the syncpoint would need to be put in a place where the mutex is held).

All of this makes me wonder if this test isn't worth checking in at all: maybe the value was just in demonstrating the bug and that it was fixed, but it's not so useful as a regression test since it's testing a very specific concurrency bug?

File: source/extensions/transport_sockets/tls/ssl_socket.cc [hunk 3]
```diff
@@ -380,13 +380,38 @@ bool ClientSslSocketFactory::implementsSecureTransport() const { return true; }
...
   {
     absl::WriterMutexLock l(&ssl_ctx_mu_);
     ssl_ctx_ = manager_.createSslClientContext(stats_scope_, *config_, ssl_ctx_);
+    if (ssl_ctx_) {
+      should_run_callbacks = true;
+    }
+  }
+  if (should_run_callbacks) {
+    for (const auto& cb : secrets_ready_callbacks_) {
+      cb();
```
Comment (antoniovicente) [inline source/extensions/transport_sockets/tls/ssl_socket.cc:393]: On what thread to these callbacks run?

I don't know what the threading model looks like for health checker vs SDS.  I think there's potential for callbacks being invoked from the wrong thread.

File: source/extensions/transport_sockets/tls/ssl_socket.cc [hunk 4]
```diff
@@ -380,13 +380,38 @@ bool ClientSslSocketFactory::implementsSecureTransport() const { return true; }
...
+  bool should_run_callbacks = false;
   {
     absl::WriterMutexLock l(&ssl_ctx_mu_);
     ssl_ctx_ = manager_.createSslClientContext(stats_scope_, *config_, ssl_ctx_);
+    if (ssl_ctx_) {
+      should_run_callbacks = true;
+    }
+  }
+  if (should_run_callbacks) {
+    for (const auto& cb : secrets_ready_callbacks_) {
```
Comment (antoniovicente) [inline source/extensions/transport_sockets/tls/ssl_socket.cc:392]: From what threads are callbacks added to this list?  Do adds happen from outside the thread where onAddOrUpdateSecret() is called?

**Review by antoniovicente — COMMENTED** [review]
Thanks for the fixes to SDS handling and health checks.

I'ld like to better understand threading involved in SDS updates vs health checkers.  Please see comments below.

**Comment by mpuncel** [comment]
thanks for the feedback, I'll hopefully be able to address it in a couple days
